### ISSUE TITLE
Python future interacts badly with stub lifetime

### ISSUE THREAD

**Issue by alamaison** [grpc/grpc#5644]
We have problems using Python futures outside of the scope within which they are created.  For example, the following code produces a LocalShutdowError:
```python
class TestServiceImpl(test_proto.BetaTestServiceServicer):
    def GetTest(self, request, _):
        return test_proto.GetTestResponse(name=request.name)

def function(channel, timeout):
    stub = test_proto.beta_create_TestService_stub(channel)
    return stub.GetTest.future(test_proto.GetTestRequest(name='bob'), timeout)
    
response = function(self.channel, 10)
name = response.result().name
```
```
Traceback (most recent call last):
  File "grpc_client_test.py", line 43, in test_future
    name = response.result().name
  File "./python/grpc/framework/crust/_control.py", line 324, in result
    raise self._termination.abortion_error  # pylint: disable=raising-bad-type
LocalShutdownError: LocalShutdownError(code=None, details="None")
```
It looks like the Future is not maintaining a reference to the stub, so the stub is destroyed prematurely.

We attempted to work around this with the following wrapper that links stub lifetime with future lifetime:
```python
def _clean_up_stub(stub):
      del stub

class GrpcFuture(Future):
    def __init__(self, stub, future):
        self.stub = stub
        self.future = future
        self.future.add_done_callback(lambda f: _clean_up_stub(stub))

    def done(self):
        return self.future.done()

    def exception(self, timeout=None):
        return self.future.exception(timeout)

    # Passthrough etc. ....
```
That worked, and the future now returns a result outside of the scope where the stub was created.  However, it causes the interpreter to hang at exit.  If we remove the callback to clean up the stub, it doesn't hang.  Strangely, the very act of passing the stub to the clean up method seems to make the interpreter hang.

Hangs:
```python
def _clean_up_stub(stub):
      del stub
...
        self.future.add_done_callback(lambda f: _clean_up_stub(stub))
```

Hangs:
```python
def _clean_up_stub(stub):
      pass
...
        self.future.add_done_callback(lambda f: _clean_up_stub(stub))
```

Doesn't hang:
```python
def _clean_up_stub():
      pass
...
        self.future.add_done_callback(lambda f: _clean_up_stub())
```

Doesn't hang:
```python
def _clean_up_stub(future):
      pass
...
        self.future.add_done_callback(lambda f: _clean_up_stub(future))
```

Is this a bug in gRPC?

**Comment by alamaison** [comment]
CC @csdigi

**Comment by alamaison** [comment]
Forgot to mention this is with 0.12

**Comment by soltanmm** [comment]
I believe this is a reference cycle being formed among classes with `__del__` methods (a lot of gRPC Python's classes). The Python GC never collects these objects (because they have `__del__` methods and the only way to handle cycles is to arbitrarily break cycles thus making `__del__` invocations a Bad Idea™), thus underlying queues in gRPC core are never pumped, thus the program never finishes.

You can avoid this by using the explicit `with` form of stubs.

... Though I'm now of the opinion that we shouldn't even be allowing this kind of construction to begin with...

**Comment by alamaison** [comment]
> You can avoid this by using the explicit with form of stubs
It's not clear how to use context management properly with futures.  A `while` block around `GetTest.future` results in `LocalShutdownError` again (unsurprising because that destroys the stub before the future has finished).  Explicitly calling `stub.__exit__()` in a callback registered with `add_done_callback` results in the same hang as before because, again, a reference to the stub is used in the callback.

**Comment by soltanmm** [comment]
One idea is to queue up your finished event in some other Python object and to use that object to exit the stub (externalizing the reference). You could alternatively make a [weak reference](https://docs.python.org/2/library/weakref.html) to the stub and use that in your callback instead (n.b. I've not tried this particular construction yet).

**Comment by hsaliak** [comment]
Probably fixed with the recent refactoring changes. 
@alamaison would you be able to verify?
Assigning to @kpayson64

**Comment by alamaison** [comment]
@hsaliak Confirmed this is fixed in latest master

**Comment by kpayson64** [comment]
I've also verified the fix.  The fix will be included in the 0.15 release.
